talkspace/chatGPT.js

```diff
@@ -0,0 +1,295 @@
+    },
+  },
+});
+
+// Define the Booking model
+const Booking = sequelize.define('Booking', {
+  time: {
+    type: DataTypes.DATE,
+    allowNull: false,
+    validate: {
+      isDate: true,
+      isAfter: new Date().toISOString(),
+    },
+  },
+  patient: {
+    type: DataTypes.STRING,
+    allowNull: true, // Allow anonymous bookings
+  },
+  provider: {
+    type: DataTypes.STRING,
+    allowNull: false,
+  },
+  status: {
+    type: DataTypes.STRING,
+    allowNull: false,
+    defaultValue: 'pending', // Default status
+  },
+});
+
+const BookingStatusHistory = sequelize.define('BookingStatusHistory', {
+  status: {
+    type: DataTypes.STRING,
+    allowNull: false,
+  },
+  timestamp: {
+    type: DataTypes.DATE,
+    allowNull: false,
+    defaultValue: Sequelize.literal('CURRENT_TIMESTAMP'),
+  },
+});
+
+// Define the relationship between Booking and Credit
+Booking.belongsTo(Credit);
+Credit.hasOne(Booking);
+
+// Define the relationship between Booking and BookingStatusHistory
+Booking.hasMany(BookingStatusHistory);
+BookingStatusHistory.belongsTo(Booking);
+
+// Function to execute database queries
+async function executeQuery(query: string, params: any[] = []) {
+  const connection = await sequelize.getConnection();
+  try {
+    const [rows] = await connection.query(query, params);
+    return rows;
+  } catch (error) {
+    throw error;
+  } finally {
+    connection.release();
+  }
+}
+
+// Function to get statistics on canceled and rescheduled bookings for a specific provider
+async function getStats(providerId) {
+  try {
+    // Retrieve canceled and rescheduled bookings for the specified provider
+    const stats = await Booking.findAll({
+      attributes: [
+        [sequelize.fn('COUNT', sequelize.literal('DISTINCT CASE WHEN status = "canceled" THEN id END')), 'canceledBookings'],
+        [sequelize.fn('COUNT', sequelize.literal('DISTINCT CASE WHEN status = "rescheduled" THEN id END')), 'rescheduledBookings'],
+      ],
+      where: {
+        provider: providerId,
+        [sequelize.Op.or]: [{ status: 'canceled' }, { status: 'rescheduled' }],
+      },
+    });
+
+    // Extract the results from the stats
+    const [result] = stats;
+
+    // Prepare the statistic information
+    const canceledBookings = result.getDataValue('canceledBookings') || 0;
+    const rescheduledBookings = result.getDataValue('rescheduledBookings') || 0;
+
+    return [
+      canceledBookings,
+      rescheduledBookings,
+    ];
+  } catch (error) {
+    console.error('Error getting cancellation and reschedule statistics:', error);
+    throw error;
+  }
+}
+
+// Function to get monthly statistics on credits used by a specific patient, including the percentage
+async function getCreditsUsedStats(patientId) {
+  try {
+    // Retrieve total credits available for the specified patient
+    const totalCreditsQuery = await Credit.sum('type', {
+      where: {
+        BookingId: null, // Credits not associated with any booking
```

Comment:
Please add const 
```
const freeCreditsQuery = { where: { bookingId: null }}
```
And replace this comment with that variable.
Also, BookingId is not defined please add it as well to the definition